hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -112,13 +113,59 @@ private void initIfNeeded() {
 
   @Override
   protected Option<HoodieRecord<HoodieMetadataPayload>> getRecordByKeyFromMetadata(String key) {
+    // This function can be called in parallel through multiple threads. For each thread, we determine the thread-local
+    // versions of the baseFile and logRecord readers to use.
+    // - If reuse is enabled, we use the same readers and dont close them
+    // - if reuse is disabled, we open new readers in each thread and close them
+    HoodieFileReader localFileReader = null;
+    HoodieMetadataMergedLogRecordScanner localLogRecordScanner = null;
+    synchronized (this) {
+      if (!metadataConfig.enableReuse()) {
+        // reuse is disabled so always open new readers
+        try {
+          Pair<HoodieFileReader, HoodieMetadataMergedLogRecordScanner> readers = openReaders();
+          localFileReader = readers.getKey();
+          localLogRecordScanner = readers.getValue();
+        } catch (IOException e) {
+          throw new HoodieIOException("Error opening readers", e);
+        }
+      } else if (baseFileReader == null && logRecordScanner == null) {
+        // reuse is enabled but we haven't opened the readers yet
+        try {
+          Pair<HoodieFileReader, HoodieMetadataMergedLogRecordScanner> readers = openReaders();
+          localFileReader = readers.getKey();
+          localLogRecordScanner = readers.getValue();
+          // cache the readers
+          baseFileReader = localFileReader;
+          logRecordScanner = localLogRecordScanner;
+        } catch (IOException e) {
+          throw new HoodieIOException("Error opening readers", e);
+        }
+      } else {
+        // reuse the already open readers
+        ValidationUtils.checkState((baseFileReader != null || logRecordScanner != null), "Readers should already be open");
+        localFileReader = baseFileReader;
+        localLogRecordScanner = logRecordScanner;
+      }
+    }
```

Comment:
Do you think this might simplify things.
```
Pair<HoodieFileReader, HoodieMetadataMergedLogRecordScanner> readers = null;
if( ! reuse)
   readers = openReaders();
else if(cached readers are null)
   readers = openReaders();
   // update cached/instance variables for readers
else
  readers = Pair.of(cached base file reader, cached log file reader) 
```

Reply:
I suggest, we just stick to the instance variables and initialize/close them lazily as needed. This is typically the OO pattern

Reply:
@vinothchandar We cannot use the instance variables across multiple threads when resuse=False. 

Consider the two cases separately:

---------- reuse=FALSE
1. Each thread should open and close their own copy of readers
2. We cannot use instance variables 

--------- reuse=TRUE
1. Each thread should share the same readers
2. The first thread should open the readers and initialize the instance variables

Reply:
@vinothchandar : Can we sync up on this sometime and get a closure. Would like to have this in before our next release. May be at the end of next week's sync meeting, we can discuss on this.

Reply:
@prashantwason and I already synced on this. Will be catching up on reviews.